clients/cobol-lsp-vscode-extension/src/services/copybook/downloader/ZoweExplorerDownloader.ts

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2024 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+import * as vscode from "vscode";
+import { DownloadUtil } from "./DownloadUtil";
+import * as iconv from "iconv-lite";
+import { Utils } from "../../util/Utils";
+import { SettingsService } from "../../Settings";
+import { CopybookURI } from "../CopybookURI";
+import {
+  DOWNLOAD_QUEUE_LOCKED_ERROR_MSG,
+  INVALID_CREDENTIALS_ERROR_MSG,
+  PROFILE_NAME_PLACEHOLDER,
+  UNLOCK_DOWNLOAD_QUEUE_MSG,
+} from "../../../constants";
+import { TelemetryService } from "../../reporter/TelemetryService";
+
+export abstract class ZoweExplorerDownloader {
+  protected explorerAPI: IApiRegisterClient | undefined;
+  private static lockedProfile: Set<string> = new Set();
+
+  constructor(private readonly storagePath: string) {
+    this.loadZEApis();
+  }
+
+  private async loadZEApis() {
+    this.explorerAPI = await Utils.getZoweExplorerAPI();
+  }
+
+  protected abstract downloadCopybookContent(
+    dataset: string,
+    member: string,
+    profileName: string,
+  ): Promise<void>;
+
+  protected getZoweProfile(profileName: string) {
+    return this.explorerAPI!.getExplorerExtenderApi()
+      .getProfilesCache()
+      .loadNamedProfile(profileName);
+  }
+
+  protected getDownloadOptions(
+    profileName: string,
+    dataset: string,
+    member: string,
+    loadedProfile: IProfileLoaded,
+  ) {
+    const downloadBinary = !!SettingsService.getCopybookFileEncoding();
+    const baseUri = vscode.Uri.file(
+      CopybookURI.createDatasetPath(profileName, dataset, this.storagePath),
+    );
+    const filePath = vscode.Uri.joinPath(
+      baseUri,
+      DownloadUtil.getFilenameWithoutExtension(member),
+    );
+
+    const downloadOptions = {
+      file: filePath.fsPath,
+      binary: downloadBinary,
+      returnEtag: true,
+    };
+
+    if (!SettingsService.getCopybookFileEncoding()) {
+      (downloadOptions as any).encoding = loadedProfile.profile.encoding;
+    }
+
+    return { downloadOptions, filePath, downloadBinary };
```

Comment:
It seems like a better idea to return `string | undefined` instead of `downloadBinary` with the desired encoding value provided by the user.

Reply:
updated